SCons/Tool/ninja.py

```diff
@@ -0,0 +1,1570 @@
+            # slice up the outputs so we have a single output which we will
+            # use for the "real" builder and multiple phony targets that
+            # match the file names of the remaining outputs. This way any
+            # build can depend on any output from any build.
+            build["outputs"].sort()
+            if rule is not None and (rule.get("deps") or rule.get("rspfile")):
+                first_output, remaining_outputs = (
+                    build["outputs"][0],
+                    build["outputs"][1:],
+                )
+
+                if remaining_outputs:
+                    ninja.build(
+                        outputs=remaining_outputs, rule="phony", implicit=first_output,
+                    )
+
+                build["outputs"] = first_output
+
+            if "inputs" in build:
+                build["inputs"].sort()
+            
+            ninja.build(**build)
+
+        template_builds = dict()
+        for template_builder in template_builders:
+
+            # Special handling for outputs and implicit since we need to
+            # aggregate not replace for each builder.
+            for agg_key in ["outputs", "implicit"]:
+                new_val = template_builds.get(agg_key, [])
+
+                # Use pop so the key is removed and so the update
+                # below will not overwrite our aggregated values.
+                cur_val = template_builder.pop(agg_key, [])
+                if isinstance(cur_val, list):
+                    new_val += cur_val
+                else:
+                    new_val.append(cur_val)
+                template_builds[agg_key] = new_val
+
+            # Collect all other keys
+            template_builds.update(template_builder)
+
+        if template_builds.get("outputs", []):
+            ninja.build(**template_builds)
+
+        # We have to glob the SCons files here to teach the ninja file
+        # how to regenerate itself. We'll never see ourselves in the
+        # DAG walk so we can't rely on action_to_ninja_build to
+        # generate this rule even though SCons should know we're
+        # dependent on SCons files.
+        #
+        # TODO: We're working on getting an API into SCons that will
+        # allow us to query the actual SConscripts used. Right now
+        # this glob method has deficiencies like skipping
+        # jstests/SConscript and being specific to the MongoDB
+        # repository layout.
+        ninja.build(
+            self.ninja_file.path,
+            rule="REGENERATE",
+            implicit=[
+                self.env.File("#SConstruct").path,
+                __file__,
+            ]
+            + sorted(glob("src/**/SConscript", recursive=True)),
+        )
+
+        # If we ever change the name/s of the rules that include
+        # compile commands (i.e. something like CC) we will need to
+        # update this build to reflect that complete list.
+        ninja.build(
```

Comment:
Which method is the most accurate? Currently the ninja db hides the files in response files so it's more difficult to unpack.

Reply:
For the expansion in compilation database, Ninja has the -x option for compdb added in this PR (I didn't see where that was actually documented): https://github.com/ninja-build/ninja/pull/1223

I think both will need to be enabled because for unhandled actions, Ninja falls back to SCons to build so for some builds, the Ninja comiledb by itself is not the complete build. I think only the ninja tool needs to detect if the compilation_db tool is in the environment, but detection is not straight forward (depends on the order the tools are initialized) and in the current BETA state, ninja tool affects all environment while the SCons compilation db tool may not.

The ninja tool could potentially then perform a final combination of the compilation data from both the SCons and Ninja compiledb output.